### PR TITLE
DevDocs: add syntax highlighting

### PR DESCRIPTION
**Pull request by spen**
This PR adds syntax highlighting to the code blocks shown in DevDocs.
Currently code blocks are wrapped in `<pre>` and `<code>` tags via the `marked` package but the content itself is not processed for syntax.
@alisterscott pointed this out in a recent PR of mine (#10329)

This is not a perfect solution however... I've been working with the [EmptyContent DevDoc page](https://wpcalypso.wordpress.com/devdocs/client/components/empty-content/README.md) while developing and noticed that some `.jsx` syntax is not parsed properly.

emptycontent-syntax

In this PR `marked` is configured to use `prismjs` to handle syntax highlighting. 
I've added some styling that is fairly close to GitHubs (at least in terms of colors).

So far I've tried a similar approach with highlight.js too, though this is also lacking in perfect `.jsx` support.

There is certainly an improvement but I wouldn't expect a merge right away until other options are explored.

I thought I'd get this up early in the hopes of starting a conversation - hopefully somebody out there has more experience than I do with highlighting packages :)

### Testing

Test various syntax types, checking that A) they are highlighted and B) they are highlighted correctly (allowing for caveats mentioned in the discussion):
- js: [ButtonGroup](http://calypso.localhost:3000/devdocs/client/components/button-group/README.md) (ignoring jsx here)
- javascript: [UrlPreview](http://calypso.localhost:3000/devdocs/client/blocks/url-preview/README.md) (ignoring jsx here)
- jsx: [EmptyContent](http://calypso.localhost:3000/devdocs/client/components/empty-content/README.md)
- es6: [SyncHandler](http://calypso.localhost:3000/devdocs/client/lib/wp/sync-handler/README.md)
- html: [MultiCheckbox](http://calypso.localhost:3000/devdocs/client/components/forms/multi-checkbox/README.md)
- json: [Extensions](http://calypso.localhost:3000/devdocs/client/extensions/README.md)

Notes: the only other syntax in use in code-blocks is 'text' (from a project-wide search for all instances of ` ``` ` ).

### CODE REVIEW

**Comment by dmsnell** [comment]
This is great! thanks @spen!

the code looks fine to me but I'd be interested in knowing why you chose prism over highlight and other options if you also looked into them

**Comment by spen** [comment]
So I only really looked in to [Prism](https://github.com/PrismJS/prism) and [Highlight.js](https://github.com/isagalaev/highlight.js) as they seemed to surface the most.
I was really hoping to find and use the exact tool that github uses to be able to just use their OSS 'theme', matching any code-blocks one to one between GitHub & Calypso's DevDocs but from what I can gather it's a ruby gem and it doesn't seem to be open source ( there's been discussion around this here: github/pages-gem#160 ). 
I'm going to look in to one or two more tonight, including Atoms ['highlights'](https://github.com/atom/highlights) - being that it's a GH product I'd hazard a guess that it's fairly close to the output on GitHub.com.
After that I'll write something a bit more detailed... until then though it's mostly down to `jsx` syntax support (but neither Highlight.js or Prism seem to have it properly dialed in 100% 😕).

**Comment by spen** [comment]
I've put together some notes on some of the open-source syntax highlighters currently available.
Feel free to chime in if there are any other good candidates that I've missed and I'll add them to this document 🙂

## Requirements
Here're a few points that I think are important to keep in mind when considering one of the projects:

- Should not be applied retrospectively to existing DOM elements (no relying on document.xx etc.) rather it should be applied to a string.
- Compatible with [`marked`](https://github.com/chjj/marked) since that's already in use.
- Good support for `.jsx` & `.es6` syntax - This should be a priority in Calypso.
  - Including JS inside of JSX inside of JS etc.
- General OSS points - is it actively maintained, is it popular, well documented etc.

## notes ...
### [Highlight.js](https://github.com/isagalaev/highlight.js)
This seems popular and active.

Unfortunately it seems to struggle somewhat with JSX syntax - particularly JS within JSX.
Take a look at this render (Using v 9.1 - see notes):
screen shot 2017-01-08 at 13 00 29
And it's output markup:
screen shot 2017-01-08 at 13 01 15
Notice that syntax in `{ things.map( ...` is not picked up on, rather left as one 'token'.

- JSX support reportedly broken after v9.1 [reference](https://github.com/isagalaev/highlight.js/issues/1383)
- Dan Abramov reports a JSX bug ([here](https://github.com/isagalaev/highlight.js/issues/931#issuecomment-139781987)) and mentions that he has switched to prism (for it's better jsx support). Read on and you'll find that this issue was reopened in April 2016 and has been referenced a number of times since.

### [Prism](https://github.com/PrismJS/prism)
This also seems popular and active - I've seen a few people mention that they've made the switch from highlight.js to this package based on the `.jsx` support (see above)

There's a tool for testing Prism's highlighting [here](http://prismjs.com/test.html).
If you use the following code example then you'll notice that there's still the slight issue with `jsx` props
prism-props

For the initial commit here in this PR I chose to use Prism because ~~of Dan Abramov~~ the slight issues with the props seems more reasonable than losing support of the JS within JSX.

### [highlights](https://github.com/atom/highlights)

This highlighter is used in Atom's [Markdown Preview](https://github.com/atom/markdown-preview).
Being that Atom is a GitHub product I would expect its syntax support to be inline with that of github.com's.
github.com seems to use a ruby gem named 'prettylights' which isn't open source, something to do with it holding a GPLv3 license (https://github.com/github/pages-gem/issues/160#issuecomment-134565796).

From what I can see, the intent is that this reads a *file* and outputs the tokenized contents.
Following the logic through, it looks like it uses the `tokenizeLines` method of [this repo/file](https://github.com/atom/first-mate/blob/master/src/grammar.coffee).

#### Negatives
- Seems to only apply to file contents
- The project seems very light on documentation
- It's also written in coffeescript 😒

### [Syntaxhighlighter](https://github.com/syntaxhighlighter/syntaxhighlighter)

I feel like I recently saw a block of code on wordpress.org that wasn't highlighted and assumed there was just no highlighting in place but when I went looking for an example I stumbled across [this update](https://make.wordpress.org/core/2016/11/07/whitespace-changes-in-navigation-for-4-7/) that includes a highlighted block.

On inspection it turns out that this uses syntaxhighlighter, applied on the client side to the DOM nodes. Quickly tested by just loading the page with JS turned off.

The (brevityified) markup for the first line looks like this:

```html
<!-- container -->
<div id="highlighter_696439" class="syntaxhighlighter  php">
<!-- omitting line numbers... -->
<code class="php plain">wp_nav_menu( </code>
<code class="php keyword">array</code>
<code class="php plain">(</code>
<!-- on so on -->
```

This approach is different to that of the other tools above in that instead of classing the containing `<code>` block with the language name and wrapping each keyword, punctuation etc. in span tags, each token is wrapped in a code block with `class="{ language } {token}"`.

I've explored syntaxhighlighter a little, looked at it's source and docs and it doesn't seem to apply to strings (I may have missed something though) and so unfortunately I'd rule that out for Calypso.

I've not gone very deep with this package - since it's applied retrospectively to the DOM so I don't know how its JSX support is.
___

Ultimately it'd be awesome to have something that can be used universally, whether thats one syntax processor that can be used all over Automattic products... or at the very least a standard output from, say, a node syntax-processor and a PHP syntax-processor so that 'theme' css can be applied in all necessary places - JSX syntax support on wordpress.org should be identical to that on Calypso's DevDocs as it should be for code-blocks in Calypso's Reader.
Different themes could of cause be used to style these areas differently but the eventual markup should be reliably standard.

**Comment by dmsnell** [comment]
Thanks @spen - this is really a valuable assessment of the options. It seems very reasonable why you chose prism.

I support this PR :smile:

File: client/devdocs/design/syntax.scss [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    -ms-hyphens: none;
+    hyphens: none;
+}
+
+code[class*="lang-"]::selection, code[class*="lang-"] ::selection,
+code[class*="lang-"]::-moz-selection, code[class*="lang-"] ::-moz-selection {
+    background: #b3d4fc;
+}
+
+.token.comment,
```
Comment (spen) [inline client/devdocs/design/syntax.scss]: I do believe there is a plug-in that allows for a custom class, which might be worth adding to better namespace these styles. I'll add that to https://github.com/Automattic/wp-calypso/pull/10359#issuecomment-271124311 .

**Comment by alisterscott** [comment]
@spen: thanks for this PR. Can you please let me know when you consider it ready for testing and I can do some testing on it.

**Comment by spen** [comment]
Sure thing! I've added some test instructions to the pull description so it could be tested now but I've not tested all of these syntaxes myself (I'm working on another project right now so can't test these myself until a little later) so you may want to wait until then.

Though I think it's ready to test I don't think it's necessarily ready to merge as I think the CSS could be improved, using variables etc.

**Comment by spen** [comment]
It seems that ` ```es6 ` and ` ```json ` are not processed by Prism.
I'll aim to fix this today - after that this should be good to test :)

**Comment by spen** [comment]
Ok, this should be good to test now (see the test instructions in the desc).

#### note: 
I've hardcoded an alias of `javascript` to `es6` so that ` ```es6 ` blocks are processed and highlighted correctly but I've also opened a pull upstream https://github.com/PrismJS/prism/pull/1079 - if/when that makes it's way to a release I'll be able to remove the alias :)

**Comment by alisterscott** [comment]
Deployed and looking great on wpcalypso: eg: https://wpcalypso.wordpress.com/devdocs/client/components/empty-content/README.md

**Comment by spen** [comment]
Ahh awesome! Thanks @alisterscott ! 
Really good to see it looking pretty live!

I think I'll do some investigating in to how those string props might be better handled and, if I can figure it out, open an upstream PR :)

File: package.json [hunk 2]
```diff
@@ -90,6 +90,7 @@
     "phone": "git+https://github.com/Automattic/node-phone.git#1.0.8",
     "photon": "2.0.0",
     "postcss-cli": "2.5.1",
+    "prismjs": "^1.6.0",
```
[the remark this answers is not in the document]
Comment (gwwar) [reply package.json]: oh sorry looks like this was already pinned elsewhere 👍

[the remark this answers is not in the document]
Comment (spen) [reply package.json]: Thanks for the nudge though @gwwar, always good to be thorough! :)

**Pull request merged**